### PR TITLE
Sarveshn/styledcomponents

### PR DESCRIPTION
**Pull request by sarveshnagpal**
Listen for insertRules call to capture cssRules for STYLE tags that do not have any textContent. This works with styled components library. The scope is restricted because performance of iterating over CSS Rules is significantly slower than grabbing textContent.

### CODE REVIEW

File: src/plugins/layout.ts [hunk 1]
```diff
@@ -39,13 +45,29 @@ export default class Layout implements IPlugin {
         subtree: true
       });
     }
+    if (this.insertRule) {
+      let that = this;
```
Comment (ssivov) [inline src/plugins/layout.ts]: >let that = this; [](start = 6, length = 16)

Use es6 arrow function - it takes care of this/that problem

File: src/plugins/layout.ts [hunk 2]
```diff
@@ -8,11 +8,15 @@ import { ShadowDom } from "./layout/shadowdom";
...
 export default class Layout implements IPlugin {
+  private readonly cssTimeoutLength = 50;
   private eventName = "Layout";
   private distanceThreshold = 5;
   private shadowDom: ShadowDom;
   private inconsistentShadowDomCount: number;
   private observer: MutationObserver;
+  private insertRule;
+  private cssTimeout;
+  private cssElementQueue;
```
Comment (ssivov) [inline src/plugins/layout.ts]: Please add typings

File: src/core.ts [hunk 3]
```diff
@@ -266,7 +266,7 @@ function init() {
   resetUploads();
 
   if (config.customInstrumentation) {
-    let customInst = config.customInstrumentation();
+    let customInst = config.customInstrumentation(impressionId, cid, config.projectId);
```
Comment (ssivov) [inline src/core.ts]: >let customInst = config.customInstrumentation(impressionId, cid, config.projectId); [](start = 4, length = 83)

Why this change?

File: src/plugins/layout.ts [hunk 4]
```diff
@@ -190,7 +273,8 @@ export default class Layout implements IPlugin {
   }
 
   private processMutation(action: Action, shadowNode: IShadowDomNode): IEventData {
-    let state = shadowNode.computeInfo().state;
+    let info = this.computeInfo(shadowNode);
+    let state = info.state;
```
Comment (ssivov) [inline src/plugins/layout.ts]: nit: state = this.computeInfo(shadowNode).state

File: src/plugins/layout.ts [hunk 5]
```diff
@@ -86,7 +108,15 @@ export default class Layout implements IPlugin {
...
   private discoverNode(node: Node): INodeInfo {
     let shadowNode = this.shadowDom.insertShadowNode(node, getNodeIndex(node.parentNode), getNodeIndex(node.nextSibling));
-    return shadowNode.computeInfo();
+    return this.computeInfo(shadowNode);
+  }
+
+  private computeInfo(shadowNode) {
+    let info = shadowNode.computeInfo();
+    if ((shadowNode.node as Element).tagName === "STYLE" && shadowNode.node.textContent.length === 0) {
+       info.state.cssRules = this.cssRules(shadowNode.node as Element);
```
Comment (ssivov) [inline src/plugins/layout.ts]: >info.state.cssRules = this.cssRules(shadowNode.node as Element); [](start = 7, length = 64)

This should be somewhere inside already existing stateprovider, so that shadownNode.computeInfo() produces complete state already, including cssRules

File: src/plugins/layout.ts [hunk 6]
```diff
@@ -86,7 +108,15 @@ export default class Layout implements IPlugin {
   // Add node to the ShadowDom to store initial adjacent node info in a layout and obtain an index
   private discoverNode(node: Node): INodeInfo {
     let shadowNode = this.shadowDom.insertShadowNode(node, getNodeIndex(node.parentNode), getNodeIndex(node.nextSibling));
-    return shadowNode.computeInfo();
+    return this.computeInfo(shadowNode);
+  }
+
+  private computeInfo(shadowNode) {
```
Comment (ssivov) [inline src/plugins/layout.ts]: >private computeInfo(shadowNode) { [](start = 2, length = 33)

Let's not create an "extra content" wrapper around computeInfo - this would de facto make output of shadowNode.computeInfo() not trustworthy

File: src/plugins/layout.ts [hunk 7]
```diff
@@ -117,13 +147,35 @@ export default class Layout implements IPlugin {
...
+      this.layoutHandler.bind(this, element, Source.Css);
+    }
+  }
+
   private layoutHandler(element: Element, source: Source) {
     let index = getNodeIndex(element);
     let nodeInfo = this.shadowDom.getNodeInfo(index);
     if (nodeInfo) {
       let layoutState = nodeInfo.state as IElementLayoutState;
+      let styleState = nodeInfo.state as IStyleLayoutState;
```
Comment (ssivov) [inline src/plugins/layout.ts]: >let styleState = nodeInfo.state as IStyleLayoutState; [](start = 6, length = 53)

declare this inside Source.CSS case

File: src/plugins/layout.ts [hunk 8]
```diff
@@ -117,13 +147,35 @@ export default class Layout implements IPlugin {
...
       switch (source) {
         case Source.Scroll:
+          layoutState = layoutState as IElementLayoutState;
```
Comment (ssivov) [inline src/plugins/layout.ts]: >layoutState = layoutState as IElementLayoutState; [](start = 10, length = 49)

Isn't layoutstate already typed as IElementLayoutState at line #174 ?

File: src/plugins/layout.ts [hunk 9]
```diff
@@ -137,19 +189,50 @@ export default class Layout implements IPlugin {
           }
           break;
         case Source.Input:
+          layoutState = layoutState as IElementLayoutState;
```
Comment (ssivov) [inline src/plugins/layout.ts]: >layoutState = layoutState as IElementLayoutState; [](start = 10, length = 49)

Same

File: src/plugins/layout.ts [hunk 10]
```diff
@@ -86,7 +108,15 @@ export default class Layout implements IPlugin {
   // Add node to the ShadowDom to store initial adjacent node info in a layout and obtain an index
   private discoverNode(node: Node): INodeInfo {
     let shadowNode = this.shadowDom.insertShadowNode(node, getNodeIndex(node.parentNode), getNodeIndex(node.nextSibling));
-    return shadowNode.computeInfo();
+    return this.computeInfo(shadowNode);
+  }
+
+  private computeInfo(shadowNode) {
+    let info = shadowNode.computeInfo();
+    if ((shadowNode.node as Element).tagName === "STYLE" && shadowNode.node.textContent.length === 0) {
```
Comment (ssivov) [inline src/plugins/layout.ts]: >shadowNode.node.textContent.length === 0 [](start = 60, length = 40)

Can non-empty styles contain extra rules? If yes and we are just ignoring that for perf reasons, please add a comment

File: src/plugins/layout/stateprovider.ts [hunk 11]
```diff
@@ -99,30 +94,6 @@ export function createElementLayoutState(element: Element, forceMask: boolean):
...
-    }
-    if (cssRules !== null) {
-      layoutState.cssRules = [];
-      for (let i = 0; i < cssRules.length; i++) {
-        layoutState.cssRules.push(cssRules[i].cssText);
-      }
-    }
-  }
-  return layoutState;
-}
```
Comment (ssivov) [inline src/plugins/layout/stateprovider.ts]: Somewhat redundant comment, but why remove this? If we need to compute cssRules (as an extra vs. default element node), we should have a function for it here. You can define the cssRules function in this file, so that it can be exported to layout.ts as well

File: src/config.ts [hunk 12]
```diff
@@ -15,7 +15,7 @@ export let config: IConfig = {
   sensitiveAttributes: [],
   timeToYield: 50,
   instrument: false,
-  cssRules: false,
+  skipCookie: false,
```
Comment (toby-walker) [inline src/config.ts]: skipCookie -> disableCookie

File: src/plugins/layout/nodeinfo.ts [hunk 13]
```diff
@@ -27,7 +26,8 @@ export function shouldIgnoreNode(node: Node, parentIgnored: boolean): boolean {
         break;
       case Node.TEXT_NODE:
         // If we capture CSSRules on style elements, ignore its text children nodes
-        if (config.cssRules && node.parentNode && (node.parentNode as Element).tagName === "STYLE") {
+        let parent = node.parentNode as Element;
+        if (parent && parent.tagName === "STYLE" && parent.textContent.length === 0) {
```
Comment (toby-walker) [inline src/plugins/layout/nodeinfo.ts]: >tagName [](start = 29, length = 7)

What is this logic doing? add comment.

File: types/config.d.ts [hunk 14]
```diff
@@ -42,16 +42,16 @@ export interface IConfig {
...
+    skipCookie?: boolean;
 
     // Pointer to the function which would be responsible for sending the data
     // If left unspecified, raw payloads will be uploaded to the uploadUrl endpoint
     uploadHandler?: UploadHandler;
 
     // Pointer to the function which will be responsible for giving Clarity
     // a dictionary of strings that the user wants logged in each Clarity payload
-    customInstrumentation?: () => { [key: string]: string; };
+    customInstrumentation?: (impressionId?: string, cid?: string, projectId?: string) => { [key: string]: string; };
```
Comment (toby-walker) [inline types/config.d.ts]: Are we restricting to a fixed set?

File: types/layout.d.ts [hunk 15]
```diff
@@ -29,7 +29,8 @@ export const enum Source {
   Discover,
   Mutation,
   Scroll,
-  Input
+  Input,
+  Css
```
Comment (toby-walker) [inline types/layout.d.ts]: >Css [](start = 2, length = 3)

Add trailing ,

File: src/plugins/layout.ts [hunk 16]
```diff
@@ -39,13 +45,29 @@ export default class Layout implements IPlugin {
         subtree: true
       });
     }
+    if (this.insertRule) {
+      let that = this;
+      CSSStyleSheet.prototype.insertRule = function(style, index) {
```
[the remark this answers is not in the document]
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Done

---
In reply to: [243139227](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243139227) [](ancestors = 243139227)

File: src/plugins/layout.ts [hunk 17]
```diff
@@ -39,13 +45,29 @@ export default class Layout implements IPlugin {
...
       this.observer.disconnect();
     }
 
+    if (this.cssTimeout) {
+      clearTimeout(this.cssTimeout);
+    }
+
+    // Restore original insertRule definition
+    CSSStyleSheet.prototype.insertRule = this.insertRule;
+    this.insertRule = null;
```
[the remark this answers is not in the document]
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Done

---
In reply to: [243139386](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243139386) [](ancestors = 243139386)

File: src/core.ts (hunk 3 above)
Comment (sarveshnagpal) [reply src/core.ts]: This is to allow webmasters to integrate Clarity into their own systems or analytics solution. E.g. a partner site can look up these values, store somewhere, and directly reference Clarity visualization. 

---
In reply to: [243067042](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243067042) [](ancestors = 243067042)

File: src/plugins/layout.ts (hunk 8 above)
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Fixed

---
In reply to: [243076492](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243076492) [](ancestors = 243076492)

File: src/plugins/layout.ts (hunk 7 above)
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Done

---
In reply to: [243076202](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243076202) [](ancestors = 243076202)

File: src/plugins/layout.ts [hunk 18]
```diff
@@ -117,13 +147,35 @@ export default class Layout implements IPlugin {
...
+    if (this.cssElementQueue.indexOf(element) === -1) {
+      this.cssElementQueue.push(element);
+    }
+
+    this.cssTimeout = setTimeout(this.cssDequeue.bind(this), this.cssTimeoutLength);
+  }
+
+  private cssDequeue() {
+    for (let element of this.cssElementQueue) {
+      this.layoutHandler.bind(this, element, Source.Css);
```
[the remark this answers is not in the document]
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: This is a signal to layoutHandler that the source of this change is coming from Css rules. Which is defined as an enum, like others.

---
In reply to: [243139943](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243139943) [](ancestors = 243139943)

File: src/plugins/layout.ts (hunk 9 above)
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Done

---
In reply to: [243076527](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243076527) [](ancestors = 243076527)

File: src/plugins/layout/nodeinfo.ts (hunk 13 above)
Comment (sarveshnagpal) [reply src/plugins/layout/nodeinfo.ts]: Added some more comments although the comment above is still accurate.

---
In reply to: [243140382](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243140382) [](ancestors = 243140382)

File: types/config.d.ts (hunk 14 above)
Comment (sarveshnagpal) [reply types/config.d.ts]: Not sure I understand the comment; The parameters are all optional (defined with a "?"). So we don't have to enforce. 

---
In reply to: [243140487](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243140487) [](ancestors = 243140487)

File: types/layout.d.ts (hunk 15 above)
Comment (sarveshnagpal) [reply types/layout.d.ts]: The last entry in a Javascript object shouldn't have a trailing ",". See Line 36 below for an example. 

---
In reply to: [243140562](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243140562) [](ancestors = 243140562)

File: src/plugins/layout.ts (hunk 6 above)
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Agreed. While coding I came full circle and now extra content wrapper is no longer needed. Fixed it.

---
In reply to: [243075614](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243075614) [](ancestors = 243075614)

[the remark this answers is not in the document]
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Refactored code. No longer applicable.

---
In reply to: [243139803](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243139803) [](ancestors = 243139803)

File: src/plugins/layout.ts (hunk 10 above)
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Refactored and this comment no longer applies.

---
In reply to: [243077973](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243077973) [](ancestors = 243077973)

[the remark this answers is not in the document]
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Refactored code and this no longer applies.

---
In reply to: [243139743](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243139743) [](ancestors = 243139743)

File: src/plugins/layout.ts (hunk 5 above)
Comment (sarveshnagpal) [reply src/plugins/layout.ts]: Agreed and done.

---
In reply to: [243075178](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243075178) [](ancestors = 243075178)

File: src/plugins/layout/stateprovider.ts (hunk 11 above)
Comment (sarveshnagpal) [reply src/plugins/layout/stateprovider.ts]: Fixed

---
In reply to: [243079814](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243079814) [](ancestors = 243079814)

File: src/config.ts (hunk 12 above)
Comment (sarveshnagpal) [reply src/config.ts]: Looks like Slava beat me to this change. He called it allowIdCookie; if we need to change it we can. 

---
In reply to: [243138961](https://github.com/Microsoft/clarity-js/pull/116#discussion_r243138961) [](ancestors = 243138961)

File: src/core.ts (hunk 3 above)
Comment (ssivov) [reply src/core.ts]: Moving conversation to a new comment in config.d.ts

---
In reply to: [245071226](https://github.com/Microsoft/clarity-js/pull/116#discussion_r245071226) [](ancestors = 245071226,243067042)

File: types/config.d.ts [hunk 19]
```diff
@@ -42,16 +45,17 @@ export interface IConfig {
...
     uploadHandler?: UploadHandler;
 
+    // XHLHttpRequest headers to be added to every upload request with the default upload handler
+    // Object is a map from header names to header values
+    uploadHeaders?: { [key: string]: string; };
+
     // Pointer to the function which will be responsible for giving Clarity
     // a dictionary of strings that the user wants logged in each Clarity payload
-    customInstrumentation?: () => { [key: string]: string; };
+    customInstrumentation?: (impressionId?: string, cid?: string, projectId?: string) => { [key: string]: string; };
```
Comment (ssivov) [inline types/config.d.ts]: >customInstrumentation?: (impressionId?: string, cid?: string, projectId?: string) => { [key: string]: string; }; [](start = 4, length = 112)

I'm not sure I fully understand. Custom instrumentation is a webmaster-defined function that returns a dictionary. Since we call config.customInstrumentation() ourselves, we can *always* provide arguments above and it's up to webmasters whether to use them and return in a customInst dictionary or not, so these args should not be optional. 

You'd also need to update config.customInstrumentation() call in core.ts to pass these arguments.

File: types/config.d.ts (hunk 14 above)
Comment (toby-walker) [reply types/config.d.ts]: Are our custom instrumentation limited to these specific properties listed here (impressionId, etc.)?

---
In reply to: [245074855](https://github.com/Microsoft/clarity-js/pull/116#discussion_r245074855) [](ancestors = 245074855,243140487)

File: src/core.ts [hunk 20]
```diff
@@ -243,14 +243,15 @@ function uploadPendingEvents() {
...
-    setCookie(Cookie, guid(), 7 * 52 * 2);
-  }
-  cid = getCookie(Cookie);
+  let cidCookie = getCookie(Cookie);
+  cid = cidCookie || guid();
   impressionId = guid();
 
+  // Set 2-year CID cookie, if allowed by config
+  if (!cidCookie && config.allowIdCookie) {
+    setCookie(Cookie, cid, 7 * 52 * 2);
```
Comment (toby-walker) [inline src/core.ts]: >7 * 52 * 2 [](start = 27, length = 10)

Minor but make this a configuration and just use a null or zero value to indicate no cookies.

File: src/plugins/layout.ts [hunk 21]
```diff
@@ -39,13 +45,36 @@ export default class Layout implements IPlugin {
...
       });
     }
+    if (this.insertRule) {
+      let that = this;
+
+      // Some popular open source libraries, like styled-components, optimize performance
+      // by injecting CSS using insertRule API vs. appending text node. A side effect of
+      // using javascript API is that it doesn't trigger DOM mutation and therefore we
+      // need to override the insertRule API and listen for changes manually.
+      CSSStyleSheet.prototype.insertRule = function(style, index) {
```
Comment (toby-walker) [inline src/plugins/layout.ts]: >CSSStyleSheet [](start = 6, length = 13)

Is this always available? Should we not detect this is supported? And also wrap in a try/catch block as this is a browser compat area.

File: types/config.d.ts (hunk 14 above)
Comment (sarveshnagpal) [reply types/config.d.ts]: For now yes; although I get your point. Let me turn this into a properties object that can include more or less features going forward. 

---
In reply to: [245143218](https://github.com/Microsoft/clarity-js/pull/116#discussion_r245143218) [](ancestors = 245143218,245074855,243140487)

File: types/config.d.ts (hunk 19 above)
Comment (sarveshnagpal) [reply types/config.d.ts]: Agree with your first comment. GOes in line with Toby's comment. I will refactor. 

---
In reply to: [245132409](https://github.com/Microsoft/clarity-js/pull/116#discussion_r245132409) [](ancestors = 245132409)

Comment (sarveshnagpal) [reply types/config.d.ts]: @Slava - The second part was already done.

---
In reply to: [245168785](https://github.com/Microsoft/clarity-js/pull/116#discussion_r245168785) [](ancestors = 245168785,245132409)

Comment (ssivov) [reply types/config.d.ts]: Ah, right, i looked at the wrong line, sorry.

---
In reply to: [245169815](https://github.com/Microsoft/clarity-js/pull/116#discussion_r245169815) [](ancestors = 245169815,245168785,245132409)
